tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java

```diff
@@ -229,37 +237,52 @@ else if ( mirrors.size() == 2 ) {
 
 	/**
 	 * @param annotations list of annotation mirrors.
-	 *
 	 * @return target entity class name as string or {@code null} if no targetEntity is here or if equals to void
 	 */
 	private String getTargetEntity(List<? extends AnnotationMirror> annotations) {
 		String fullyQualifiedTargetEntityName = null;
-		for ( AnnotationMirror mirror : annotations ) {
-			if ( TypeUtils.isAnnotationMirrorOfType( mirror, Constants.ELEMENT_COLLECTION ) ) {
-				fullyQualifiedTargetEntityName = getFullyQualifiedClassNameOfTargetEntity( mirror, "targetClass" );
+		for (AnnotationMirror mirror : annotations) {
+			if (TypeUtils.isAnnotationMirrorOfType(mirror, Constants.ELEMENT_COLLECTION)) {
+				fullyQualifiedTargetEntityName = getFullyQualifiedClassNameOfTargetEntity(mirror, "targetClass");
 			}
-			else if ( TypeUtils.isAnnotationMirrorOfType( mirror, Constants.ONE_TO_MANY )
-					|| TypeUtils.isAnnotationMirrorOfType( mirror, Constants.MANY_TO_MANY )
-					|| TypeUtils.isAnnotationMirrorOfType( mirror, Constants.MANY_TO_ONE )
-					|| TypeUtils.isAnnotationMirrorOfType( mirror, Constants.ONE_TO_ONE ) ) {
-				fullyQualifiedTargetEntityName = getFullyQualifiedClassNameOfTargetEntity( mirror, "targetEntity" );
+			else if (TypeUtils.isAnnotationMirrorOfType(mirror, Constants.ONE_TO_MANY)
+					|| TypeUtils.isAnnotationMirrorOfType(mirror, Constants.MANY_TO_MANY)
+					|| TypeUtils.isAnnotationMirrorOfType(mirror, Constants.MANY_TO_ONE)
+					|| TypeUtils.isAnnotationMirrorOfType(mirror, Constants.ONE_TO_ONE)
+					|| TypeUtils.isAnnotationMirrorOfType(mirror, ORG_HIBERNATE_ANNOTATIONS_MANY_TO_ANY)) {
+				fullyQualifiedTargetEntityName = getFullyQualifiedClassNameOfTargetEntity(mirror, "targetEntity");
 			}
-			else if ( TypeUtils.isAnnotationMirrorOfType( mirror, ORG_HIBERNATE_ANNOTATIONS_TARGET ) ) {
-				fullyQualifiedTargetEntityName = getFullyQualifiedClassNameOfTargetEntity( mirror, "value" );
+			else if (TypeUtils.isAnnotationMirrorOfType(mirror, ORG_HIBERNATE_ANNOTATIONS_TARGET)) {
+				fullyQualifiedTargetEntityName = getFullyQualifiedClassNameOfTargetEntity(mirror, "value");
 			}
 		}
 		return fullyQualifiedTargetEntityName;
 	}
 
+	private boolean isCollectionAttribute(List<? extends AnnotationMirror> annotations) {
```

Comment:
Why did you check OneToOne and ManyToOne which are not associated with collections? And also Target.